tests/cases/unittests/session.ts

```diff
@@ -0,0 +1,459 @@
+
+			expect(session.executeCommand({
+				seq: 0,
+				type: 'command',
+				command: session.customHandler
+			})).to.deep.equal({
+				response: undefined,
+				responseRequired: true
+			});
+		});
+		it('has access to the project service', () => {
+			let ServiceSession = class extends TestSession {
+				constructor() {
+					super();
+					assert(this.projectService);
+					expect(this.projectService).to.be.instanceOf(ProjectService);
+				}
+			};
+			new ServiceSession();
+		});
+	});
+
+	describe('an example of using the Session API to create an in-process server', () => {
+		let inProcHost: ServerHost = {
+			args: [],
+			newLine: '\n',
+			useCaseSensitiveFileNames: true,
+			write: (s) => lastWrittenToHost = s,
+			readFile: () => void 0,
+			writeFile: () => void 0,
+			resolvePath: () => void 0,
+			fileExists: () => false,
+			directoryExists: () => false,
+			createDirectory: () => void 0,
+			getExecutingFilePath: () => void 0,
+			getCurrentDirectory: () => void 0,
+			readDirectory: () => void 0,
+			exit: () => void 0
+		},
+		InProcSession = class extends Session {
+			private queue: protocol.Request[] = [];
+			constructor(private client: {handle: (msg: protocol.Message) => void}) {
+				super(inProcHost, Buffer.byteLength, process.hrtime, mockLogger);
+				this.addProtocolHandler('echo', (req: protocol.Request) => ({
+					response: req.arguments,
+					responseRequired: true
+				}));
+			}
+			
+			send(msg: protocol.Message) {
+				this.client.handle(msg);
+			}
+
+			enqueue(msg: protocol.Request) {
+				this.queue = [msg].concat(this.queue);
+			}
+			
+			handleRequest(msg: protocol.Request) {
+				let response: protocol.Response;
+				try {
+					response = this.executeCommand(msg).response;
+				} catch (e) {
+					this.output(undefined, msg.command, msg.seq, e.toString());
+					return;
+				}
+				if (response) {
+					this.output(response, msg.command, msg.seq);
+				}
+			}
+
+			consumeQueue() {
+				while (this.queue.length > 0) {
+					let elem = this.queue[this.queue.length - 1];
+					this.queue = this.queue.slice(0, this.queue.length - 1);
+					this.handleRequest(elem);
+				}
+			}
+		},
+		InProcClient = class {
+			private server: Session & {enqueue: (msg: protocol.Request) => void};
+			private seq: number = 0;
+			private callbacks: ts.Map<(resp: protocol.Response) => void> = {};
+			private eventHandlers: ts.Map<(args: any) => void> = {};
+
+			handle(msg: protocol.Message): void {
+				if (msg.type === 'response') {
+					let response = <protocol.Response>msg;
+					if (this.callbacks[response.request_seq]) {
+						this.callbacks[response.request_seq](response);
+						delete this.callbacks[response.request_seq];
+					}
+				} else if (msg.type === 'event') {
+					let event = <protocol.Event>msg;
+					this.emit(event.event, event.body);
+				}
+			}
+
+			emit(name: string, args: any): void {
+				if (!this.eventHandlers[name]) {
+					return;
+				}
+				this.eventHandlers[name](args);

```

Comment:
Will do. It's my habit to use fail fast if statements to reduce nesting - in this case there's no conditions to nest so it didn't really matter.